### PR TITLE
zebra-cli init

### PR DESCRIPTION
**Pull request by dconnolly**
## Motivation

Needed skeleton program for generating and displaying unified addresses

### Specifications

Pulls in `zcash_address` to encode unified addresses:
https://github.com/zcash/zips/blob/main/zip-0316.rst

### Designs

https://zebra.zfnd.org/dev/rfcs/0009-zebra-client.html

## Solution

Create `zebra-cli` with Abscissa (same framework as `zebrad`), create a subcommand `get-new-address`, which generates new Sapling and Orchard key trees and encodes them and renders them as a QR code in std out. Does not save any keys, everything is discarded.

Also has the standard `version` and `help` commands.

image

## Review

Anyone can review

## Follow Up Work

Everything else 😹

### CODE REVIEW

File: zebra-cli/Cargo.toml
Comment (teor2345) [inline zebra-cli/Cargo.toml:16]: Duplicate dependencies: please use the same commit as `zebra/Cargo.toml` (and other Zebra crates).

```suggestion
zcash_address = { git = "https://github.com/ZcashFoundation/librustzcash.git", tag = "0.5.1-zebra-v1.0.0-beta.4" }
```

File: deny.toml
Comment (teor2345) [inline deny.toml:86]: Duplicate dependencies: this duplicate should disappear if we use the same commit as `zebra/Cargo.toml`.

File: zebra-cli/Cargo.toml
Comment (teor2345) [inline zebra-cli/Cargo.toml:25]: Nitpick: Consistency with other version specifiers.

```suggestion
abscissa_core = "0.5.2"
```

File: zebra-cli/README.md
Comment (teor2345) [inline zebra-cli/README.md:3]: Missing documentation: We should explain what the application does.

```suggestion
# zebra-cli

`zebra-cli` is a command-line application for (TODO).
```

File: zebra-cli/src/commands/getnewaddress.rs
Comment (teor2345) [inline zebra-cli/src/commands/getnewaddress.rs:47]: Config: We either need to support testnet, or document that the only supported network is mainnet.

File: zebra-cli/src/commands/getnewaddress.rs
Comment (teor2345) [inline zebra-cli/src/commands/getnewaddress.rs:48]: Missing Transparent: We either need to support transparent addresses, or document that they are not supported.

File: zebra-cli/src/config.rs
Comment (teor2345) [inline zebra-cli/src/config.rs:14]: Example Code: This file contains example code, we should either delete it, or replace it with a useful config.

File: zebra-cli/src/bin/zebra-cli/main.rs
Comment (teor2345) [inline zebra-cli/src/bin/zebra-cli/main.rs:4]: Clippy: Can we use Zebra's standard clippy settings here?

File: zebra-cli/src/lib.rs
Comment (teor2345) [inline zebra-cli/src/lib.rs:16]: Clippy: Can we use Zebra's standard clippy settings here?

File: zebra-cli/src/lib.rs
Comment (teor2345) [inline zebra-cli/src/lib.rs:7]: Nitpick: Remove redundant hints.

```suggestion
```

File: zebra-cli/src/lib.rs
Comment (teor2345) [inline zebra-cli/src/lib.rs:5]: Missing App-Specific Documentation

File: zebra-cli/tests/acceptance.rs
Comment (teor2345) [inline zebra-cli/tests/acceptance.rs:19]: Clippy: Can we use Zebra's standard clippy settings here?

File: zebra-cli/tests/acceptance.rs
Comment (teor2345) [inline zebra-cli/tests/acceptance.rs:10]: Nitpick: Remove redundant hints.
```suggestion
```

File: zebra-cli/tests/acceptance.rs
Comment (teor2345) [inline zebra-cli/tests/acceptance.rs:8]: Missing Test-Specific Documentation

File: zebra-cli/tests/acceptance.rs
Comment (teor2345) [inline zebra-cli/tests/acceptance.rs:35]: Example Code: This file contains example code, we should either delete it, or replace it with useful tests.

**Comment by teor2345** [comment]
This looks great, neat features!

This PR adds another binary to our builds, and we're already exceeding the 10 GB GitHub cache limit. So I'd like to do some of these things before we merge this PR:
* set up a merge bot and turn off Windows and macOS builds on PRs
* stop caching Windows builds (they are not required to merge anyway)
* any other changes that reduce the cache size

As an alternative, could we add this new feature as a `zebrad` command?

I'm not quite sure how we split features between `zebrad`, `zebra-client`, and `zebra-cli`. So it might be helpful to discuss and document that before we create `zebra-cli`.

**Comment by dconnolly** [comment]
> This looks great, neat features!
> 
> This PR adds another binary to our builds, and we're already exceeding the 10 GB GitHub cache limit. So I'd like to do some of these things before we merge this PR:
> 
> * set up a merge bot and turn off Windows and macOS builds on PRs

💯

> * stop caching Windows builds (they are not required to merge anyway)

💯
> * any other changes that reduce the cache size

👍

> I'm not quite sure how we split features between `zebrad`, `zebra-client`, and `zebra-cli`. So it might be helpful to discuss and document that before we create `zebra-cli`.

The design in https://zebra.zfnd.org/dev/rfcs/0009-zebra-client.html basically has `zebra-client` as a library crate to handle 'all functionality related to a particular user's private data', which would then be used by the `zebra-cli` binary ('the wallet') and also 'a client component for zebrad that handles block chain scanning' which is used by `zebrad` when a `ViewingKey` is registered for block chain scanning.

So this PR should probably move the bulk `GetNewAddressCmd::run()` into `zebra-client`, at least to have those dependencies in there, and then `zebra-cli` depend on `zebra-client` to expose/re-export those dependencies.

But yeah this is strictly non-urgent and can wait to merge until we tidy up our CI caches 📦

**Comment by teor2345** [comment]
> The design in https://zebra.zfnd.org/dev/rfcs/0009-zebra-client.html basically has `zebra-client` as a library crate to handle 'all functionality related to a particular user's private data', which would then be used by the `zebra-cli` binary ('the wallet') and also 'a client component for zebrad that handles block chain scanning' which is used by `zebrad` when a `ViewingKey` is registered for block chain scanning.
> 
> So this PR should probably move the bulk `GetNewAddressCmd::run()` into `zebra-client`, at least to have those dependencies in there, and then `zebra-cli` depend on `zebra-client` to expose/re-export those dependencies.

That makes sense. But I think having `zebra-client` and `zebra-cli` could be a bit confusing, because the names are very similar.

So I'm wondering if we want to do some of these things:
- rename `zebra-client` to `zebra-wallet`
- put the wallet and client in the same crate, as a lib and bin target

**Comment by teor2345** [comment]
I marked this as draft, so we don't accidentally merge it.

File: zebra-chain/src/orchard/address.rs
Comment (jvff) [inline zebra-chain/src/orchard/address.rs:53]: Optional ideas:

1. It's possible to avoid the `Vec` heap allocation and resizing using a mutable slice borrow.
2. I think it might be a good idea to use `expect` here to detect any issues if we accidentally change something in the future.

```suggestion
        let mut bytes = [0u8; 43];
        let mut writer = &mut bytes[..];

        writer
            .write_all(&<[u8; 11]>::from(addr.diversifier))
            .expect("diversifier should fit in address buffer");

        writer
            .write_all(&<[u8; 32]>::from(addr.transmission_key))
            .expect("transmission should fit in address buffer");

        bytes
```

Comment (jvff) [reply zebra-chain/src/orchard/address.rs:53]: I had an idea for an alternative that doesn't need the runtime checks. It could be simpler if the `address_bytes` and `address_end` bindings are removed, but I was worried about having the number `11` repeated and having the risk of forgetting to update one of the occurrences in the future :sweat_smile: 

```suggestion
        let mut bytes = [0u8; 43];

        let address_bytes: [u8; 11] = addr.diversifier.into();
        let address_end = address_bytes.len();

        bytes[..address_end].copy_from_slice(&address_bytes);
        bytes[address_end..].copy_from_slice(&<[u8; 32]>::from(addr.transmission_key));

        bytes
```

File: zebra-chain/src/sapling/address.rs
Comment (jvff) [inline zebra-chain/src/sapling/address.rs:92]: Optional idea: this is similar to the situation above:

```suggestion
        let mut bytes = [0u8; 43];

        let address_bytes: [u8; 11] = addr.diversifier.into();
        let address_end = address_bytes.len();

        bytes[..address_end].copy_from_slice(&address_bytes);
        bytes[address_end..].copy_from_slice(&<[u8; 32]>::from(addr.transmission_key));

        bytes
```

File: zebra-cli/src/commands/getnewaddress.rs
Comment (jvff) [inline zebra-cli/src/commands/getnewaddress.rs:47]: Optional idea: not sure if this would actually help with readability, but I thought it could help to separate this into different methods:

```rust
fn run(&self) {
    let network = zebra_chain::parameters::Network::Mainnet;

    let sapling_address = self.new_sapling_address(network);
    let orchard_address = self.new_orchard_address(network);

    let zcash_address =
        self.new_unified_address(network, vec![sapling_address, orchard_address]);

    let qr_code = self.create_qr_code_image(&zcash_address);

    println!("\nNew Unified Zcash Address:");
    println!("\n{}\n", zcash_address);
    println!("\n{}\n", qr_code);
}
```

I think the `network` parameters could also be omitted if it's stored in `GetNewAddressCmd` as a configuration option, but maybe that's not necessary for now :thinking:

File: zebra-cli/src/config.rs
Comment (jvff) [inline zebra-cli/src/config.rs:14]: Should this be removed?

**Review by jvff — APPROVED** [review]
Looks good! I've only added some optional ideas and suggestions.

**Review by teor2345 — CHANGES_REQUESTED** [review]
We haven't made any decisions about the `zebra-cli` app yet. And we don't need to make those decisions any time soon.

But creating a new `abscissa` application could add a significant maintenance burden. There's a lot of template code, which we've heavily modified it for `zebrad`. `absicca` is also about to upgrade to `clap`. So we might want to avoid new apps until after we've done that upgrade in `zebrad`.

Here are some alternatives to creating a new app:
- a new `zebrad` command
- a new `zebrad`, `zebra-client`, or `zebra-chain` example (which uses a new `zcash_address` feature in the crate)

Since there's only one real function in this PR, and it only uses `zebra-chain`, adding a `zebra-chain` example seems like the easiest option for now.

We can move the QR code function into Zebra's client app later, once we've decided what it's going to look like.

(I also have a bunch of detailed comments, but I think we should decide on the code structure first.)

**Comment by conradoplg** [comment]
Should we close this for now?

**Comment by conradoplg** [comment]
Should we close this for now?

**Comment by dconnolly** [comment]
I'm going to close this PR but leave the branch up

**Pull request closed without merging**
